src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java

```diff
@@ -0,0 +1,396 @@
+                        currentLocatables.add(read);
+                    }
+                    if (nextShard != null && overlaps(nextShard, read)) {
+                        nextLocatables.add(read);
+                    }
+                }
+                // current shard is finished, either because the current read is to the right of it, or there are no more locatables
+                Tuple2<I, Iterable<R>> tuple = new Tuple2<>(currentShard, currentLocatables);
+                currentShard = nextShard;
+                nextShard = peekingShards.hasNext() ? peekingShards.next() : null;
+                currentLocatables = nextLocatables;
+                nextLocatables = Lists.newArrayList();
+                return tuple;
+            }
+        };
+        return Iterators.filter(iterator, input -> input._2().iterator().hasNext());
+    }
+
+    /**
+     * @return <code>true</code> if the read is to the right of the given interval
+     */
+    private static <I extends Locatable, R extends Locatable> boolean toRightOf(I interval, R read, SAMSequenceDictionary sequenceDictionary) {
+        int intervalContigIndex = sequenceDictionary.getSequenceIndex(interval.getContig());
+        int readContigIndex = sequenceDictionary.getSequenceIndex(read.getContig());
+        return (intervalContigIndex == readContigIndex && interval.getEnd() < read.getStart()) // read on same contig, to the right
+                || intervalContigIndex < readContigIndex; // read on subsequent contig
+    }
+
+    /**
+     * For each partition, find the interval that spans it.
+     */
+    static <R extends Locatable> List<PartitionLocatable<SimpleInterval>> computePartitionReadExtents(JavaRDD<R> locatables, SAMSequenceDictionary sequenceDictionary, int maxLocatableLength) {
+        // Find the first read in each partition. This is very efficient since only the first record in each partition is read.
+        List<R> splitPoints = locatables.mapPartitions((FlatMapFunction<Iterator<R>, R>) it -> ImmutableList.of(it.next())).collect();
+        List<PartitionLocatable<SimpleInterval>> extents = new ArrayList<>();
+        for (int i = 0; i < splitPoints.size(); i++) {
+            Locatable current = splitPoints.get(i);
+            int intervalContigIndex = sequenceDictionary.getSequenceIndex(current.getContig());
+            final Locatable next;
+            final int nextContigIndex;
+            if (i < splitPoints.size() - 1) {
+                next = splitPoints.get(i + 1);
+                nextContigIndex = sequenceDictionary.getSequenceIndex(next.getContig());
+            } else {
+                next = null;
+                nextContigIndex = sequenceDictionary.getSequences().size();
+            }
+            if (intervalContigIndex == nextContigIndex) { // same contig
+                addPartitionReadExtent(extents, i, current.getContig(), current.getStart(), next.getStart() + maxLocatableLength);
```

Comment:
So we're now overestimating the extent of each partition by `maxLocatableLength`, and this is set to 1000 by default, correct? Can you briefly explain the performance implications of this? I assume that we'll just potentially pull in a few extra partitions per interval that don't actually contain reads overlapping the interval?